contracts/dex/app/src/lib.rs

```diff
@@ -0,0 +1,424 @@
+            }
+
+            // Calculate the liquidity tokens to mint
+            let liquidity = U256::min(
+                (amount_a * storage.total_liquidity) / storage.reserve_a,
+                (amount_b * storage.total_liquidity) / storage.reserve_b,
+            );
+
+            // Update reserves and total liquidity
+            storage.reserve_a += amount_a;
+            storage.reserve_b += amount_b;
+            storage.total_liquidity += liquidity;
+
+            // Assign liquidity tokens to the user
+            let user_liquidity = storage
+                .liquidity_providers
+                .entry(sender)
+                .or_insert(U256::zero());
+            *user_liquidity += liquidity;
+        }
+
+        let request_a = vft_io::TransferFrom::encode_call(sender, program_id, amount_a);
+        msg::send_bytes_for_reply(storage.token_a, request_a, 0, 0)
+            .expect("Error in async message to vft contract")
+            .await
+            .expect("Error getting answer from the vft contract");
+
+        let request_b = vft_io::TransferFrom::encode_call(sender, program_id, amount_b);
+        msg::send_bytes_for_reply(storage.token_b, request_b, 0, 0)
+            .expect("Error in async message to vft contract")
+            .await
+            .expect("Error getting answer from the vft contract");
+
+        self.notify_on(Event::AddedLiquidity {
+            sender,
+            amount_a,
+            amount_b,
+            liquidity: storage.total_liquidity,
+        })
+        .expect("Notification Error");
+    }
+
+    /// Removes liquidity from the pool.
+    /// The amount of liquidity to remove is specified as `amount`.
+    /// Proportionally calculates the amounts of token A and B to return to the user.
+    /// Transfers the calculated token amounts back to the user.
+    /// Updates the user's liquidity balance and the reserves of the pool.
+    pub async fn remove_liquidity(&mut self, amount: U256) {
+        let storage = self.get_mut();
+        let sender = msg::source();
+        let program_id = exec::program_id();
+
+        // Fetch the user's liquidity and ensure they have enough to remove
+        let user_liquidity = storage
+            .liquidity_providers
+            .get_mut(&sender)
+            .expect("No liquidity");
+        if *user_liquidity < amount {
+            panic!("Insufficient liquidity");
+        }
+
+        let amount_a = (amount * storage.reserve_a) / storage.total_liquidity;
+        let amount_b = (amount * storage.reserve_b) / storage.total_liquidity;
+
+        storage.reserve_a -= amount_a;
+        storage.reserve_b -= amount_b;
+        storage.total_liquidity -= amount;
+        *user_liquidity -= amount;
+
+        // Transfer token A back to the user
+        let request_a = vft_io::TransferFrom::encode_call(program_id, sender, amount_a);
+        msg::send_bytes_for_reply(storage.token_a, request_a, 0, 0)
```

Comment:
The same here.
1) The transfer occurs after the state changing
2) Running out of gas is a critical issue here